toc/rfc/rfc-draft-generic-per-route-features.md

```diff
@@ -0,0 +1,88 @@
+made aware of new fields without adding any significant logic but the over-head of making the
+change still exists.
+
+On the other side are the components that actually need to be aware of the change. Mostly the
+routing components (gorouter, maybe route-emitter / -registrar), the API (cloud controller)
+and UIs (CLI).
+
+## Proposal
+
+To reduce the overall overhead of implementing new per-route features we propose to add a generic
+key-value map to the definition of a route across all components. All components handling the map
+SHOULD ignore the contents of the map except gorouter, cloud controller and cf CLI which MUST handle
+the map and its values accordingly. Other components MAY limit the size of the map or size of
+keys / values for technical reasons.
+
+### Required Changes
+
+#### CLI
+
+Support for a generic map MUST be added to the manifest. There MAY be either individual command line
+flags or a single flag which supports generic key-value pairs for the `map-route` sub-command.
+
+#### Cloud Controller
+
+The API MUST specify and implement a new field to carry the map inside the route object. The API
+MAY implement per-field validations as features of the map are specified. The route MAY be
+rejected with an appropriate error message if the provided map is invalid and MUST be passed on
+as-is otherwise. The API also MUST store the map as a generic key-value map to ensure changes to 
+the map do not require a change to the database schema.
+
+To-Do:
+* Make the validation required?
+
+#### BBS, Rep, Executor, Route-Emitter
+
+The diego components MUST add support for the new field specified by the API and forward the
+contents as-is without modification or, if the field exceeds technical limitations, MUST raise an 
+error.
+
+#### Gorouter
+
+Gorouter MUST accept the new field and parse its contents. If the map cannot be parsed it MUST 
+be ignored. If individual key-value pairs are invalid those MUST be ignored and other key-value 
+pairs MUST still be parsed and acted upon.
+
+It MAY be decided to implement one feature as part of this RFC which would be implemented together
+with the generic logic for supporting the new map.
+
+As new features are specified they MUST be implemented in gorouter.
+
+To-Do:
+* Can we somehow report errors from gorouter back to the user?
+
+### Specifying Features
+
+It's the responsibility of the App Runtime Platform Working Group (ARD-WG) to specify new 
+features based on what is specified in this RFC. When a new feature is proposed the ARD-WG uses
+the process it would use for any new feature to decide whether and how it should be implemented. It
+MAY involve other working groups in the process.
+
+### Features Which Could be Implemented
+
+So far we have collected a list of features which would benefit from a more flexible approach to 
+per-route features:
+
+- Custom Load-Balancing Algorithm
+- Custom Connection Limits
+
```

Comment:
```suggestion
- Custom session cookie name (currently set on platform level, mostly to JSESSIONID)
- Option to trim path mapping on request (i.e. an app mapped to `/some-path/` receiving a request `GET /some-path/books` will see the request as `GET /books` if the `trim-path` option is set, instead of the pull path)
```